crypto/evp/evp_enc.c

```diff
@@ -90,8 +92,6 @@ static int evp_cipher_init_internal(EVP_CIPHER_CTX *ctx,
     ENGINE *tmpimpl = NULL;
 #endif
 
-    ctx->iv_len = -1;
```

Comment:
In the event that cipher is non-NULL then shouldn't we reset the IV length cache in that case? e.g. the cipher might be different to the one that was used previously

Reply:
Also what about the case where `EVP_CIPHER_CTX_set_params` has been called to change the default IV length? If we subsequently re-init the cipher ctx we will still be using the old cached value.

Reply:
The first time EVP_CipherInit is used iv_len must be -1 because the context is fresh. Otherwise I believe this is caught by the call to EVP_CIPHER_CTX_reset when cipher is set:

https://github.com/openssl/openssl/pull/17862/files#diff-324577d2b893cdd216c26eaa03831ce61648934291a14faa0d000f938eca295aR162

There's a similar call to EVP_CIPHER_CTX_reset in the legacy path.

Reply:
@mattcaswell This is updated here, right? https://github.com/openssl/openssl/blob/65313b774dbdfd9bd90a971f47f570ba3a5bbaac/crypto/evp/evp_enc.c#L1275

Reply:
Yes. I was worried about the case where a new `ctx->algctx` gets created and the cache is different to the default. But looking at the code I'm not sure that can happen.

Reply:
I was being extra conservative by including this one, I think you're correct in that it can be safely removed but it will also require the error handling in the `EVP_CIPHER_CTX_set_params()` function to invalidate the cached items if either `OSSL_PARAM_get_int()` calls fail:

```c
    if (ctx->cipher != NULL && ctx->cipher->set_ctx_params != NULL) {
        r = ctx->cipher->set_ctx_params(ctx->algctx, params);
        if (r > 0) {
            p = OSSL_PARAM_locate_const(params, OSSL_CIPHER_PARAM_KEYLEN);
            if (p != NULL && !OSSL_PARAM_get_int(p, &ctx->key_len)) {
                r = 0;
                ctx->key_len = -1;
            }
            p = OSSL_PARAM_locate_const(params, OSSL_CIPHER_PARAM_IVLEN);
            if (p != NULL && !OSSL_PARAM_get_int(p, &ctx->iv_len)) {
                r = 0;
                ctx->iv_len = -1;
            }
        }
    }
```

My rationale here is that on a failure, it would be possible for the EVP context and the provider to have different ideas about the lengths.  Later, calling `EVP_CipherInit_ex2()` or `EVP_CipherInit_ex()` will not necessarily force the `EVP_CIPHER_CTX_reset()` to happen.

I see another deeper problem with the `EVP_CIPHER_CTX_set_params()` function.  Does a change in the key length potentially alter the IV length (and vice versa)?  I.e. should it be more aggressive about invalidate caches.  Anyway, not something that's currently relevant.

Reply:
> Does a change in the key length potentially alter the IV length (and vice versa)

Indeed, in theory *any* change in parameters might have an impact on the IV length!